File: src/worker/normalize/canonical.go
// Package normalize turns a URL into the canonical "host/path[?query]" pattern
// stored in D1, and expands a looked-up URL into its host-suffix/path-prefix
// candidate patterns.
//
// The percent-encoding, IP-literal and candidate-expansion logic is copied from
// github.com/google/safebrowsing (see sb_urls.go). Parsing is done with the
// host language's URL parser so that this file and worker/src/canonical.ts
// produce byte-identical output. Both are tested against testdata/normalize.json.
package normalize

import (
	"crypto/sha256"
	"errors"
	"net/url"
	"regexp"
	"strings"

	"golang.org/x/net/idna"
)

var (
	// ErrInvalid means the string is not a usable http(s) URL.
	ErrInvalid = errors.New("normalize: invalid url")
	// ErrIPHost means the host is an IP literal. Feeds skip these; the Worker
	// rejects them (legitimate short links do not use bare IPs).
	ErrIPHost = errors.New("normalize: ip literal host")
)

var (
	dotsRegexp          = regexp.MustCompile(`[.]+`)
	possibleIPRegexp    = regexp.MustCompile(`^(?i)((?:0x[0-9a-f]+|[0-9\.])+)$`)
	trailingSpaceRegexp = regexp.MustCompile(`^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}) `)

	// schemeRegexp captures an explicit "scheme://" prefix, so that
	// "evil.com:8080/x" is treated as schemeless rather than as scheme "evil.com".
	schemeRegexp = regexp.MustCompile(`^([a-zA-Z][a-zA-Z0-9+.-]*)://`)

	// hostRegexp must match HOST_RE in canonical.ts. It requires at least two
	// labels and a TLD starting with a letter, so numeric hosts never pass.
	hostRegexp = regexp.MustCompile(`^(?:[a-z0-9_](?:[a-z0-9_-]{0,61}[a-z0-9_])?\.)+[a-z][a-z0-9-]{0,62}$`)

	// UTS-46 non-transitional, lenient, matching the WHATWG URL parser.
	idn = idna.New(
		idna.MapForLookup(),
		idna.Transitional(false),
		idna.StrictDomainName(false),
		idna.CheckHyphens(false),
	)
)

// Parts is a parsed, canonicalized URL. Scheme, port, userinfo and fragment are
// dropped: they never affect which page is served.
type Parts struct {
	Host  string // lowercase, punycode, no trailing dot
	Path  string // canonical, always starts with "/"
	Query string // canonical, without the leading "?"
}

// Pattern is the canonical string that gets hashed and stored.
func (p Parts) Pattern() string {
	if p.Query != "" {
		return p.Host + p.Path + "?" + p.Query
	}
	return p.Host + p.Path
}

// Hash is the key stored in D1 (bad_url.h).
func Hash(pattern string) [32]byte { return sha256.Sum256([]byte(pattern)) }

// Canonical is the feed-side entry point: one URL in, one pattern out.
func Canonical(raw string) (string, error) {
	p, err := Parse(raw)
	if err != nil {
		return "", err
	}
	return p.Pattern(), nil
}

// Candidates is the lookup-side entry point: every pattern that, if present in
// the blocklist, should block raw. Up to 7 hosts x 6 paths, deduplicated.
func Candidates(p Parts) []string {
	hosts := generateLookupHosts(p.Host, false)
	paths := generateLookupPaths(p.Path, p.Query)

	out := make([]string, 0, len(hosts)*len(paths))
	seen := make(map[string]struct{}, len(hosts)*len(paths))
	for _, h := range hosts {
		for _, pa := range paths {
			c := h + pa
			if _, ok := seen[c]; ok {
				continue
			}
			seen[c] = struct{}{}
			out = append(out, c)
		}
	}
	return out
}

// Parse canonicalizes raw. Keep every step in sync with parse() in canonical.ts.
//
// This is the feed side, so it is deliberately scheme-agnostic: it accepts http
// and https alike and drops the scheme from the pattern. Feeds are full of http
// URLs, and a feed entry "http://evil.com/x" must block a user's
// "https://evil.com/x". The shortener's https-only rule is user-input policy and
// lives in parseUserUrl() in the Worker, not here.
func Parse(raw string) (Parts, error) {
	pre, ok := preprocess(raw)
	if !ok {
		return Parts{}, ErrInvalid
	}
	u, err := url.Parse(pre)
	if err != nil {
		return Parts{}, ErrInvalid
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return Parts{}, ErrInvalid
	}
	if strings.Contains(u.Host, "[") { // IPv6 literal
		return Parts{}, ErrIPHost
	}

	host, err := canonHost(u.Hostname())
	if err != nil {
		return Parts{}, err
	}

	path, err := canonPath(u.EscapedPath())
	if err != nil {
		return Parts{}, ErrInvalid
	}

	query, err := recursiveUnescape(u.RawQuery)
	if err != nil {
		return Parts{}, ErrInvalid
	}

	return Parts{Host: host, Path: path, Query: escapeQuery(query)}, nil
}

// preprocess mirrors what the WHATWG parser does before parsing, so that
// net/url sees the same string the Worker's `new URL()` sees. ok is false when
// the WHATWG host parser would have thrown.
func preprocess(raw string) (string, bool) {
	if i := strings.IndexByte(raw, '#'); i >= 0 { // drop the fragment
		raw = raw[:i]
	}
	raw = strings.TrimSpace(raw)
	raw = strings.NewReplacer("\t", "", "\r", "", "\n", "").Replace(raw)
	raw = strings.ReplaceAll(raw, `\`, "/") // special schemes treat \ as /
	raw = escapeLonePercent(raw)            // net/url rejects "%zz"; WHATWG encodes it

	m := schemeRegexp.FindStringSubmatch(raw)
	if m == nil {
		raw = "http://" + raw // some feeds ship bare "evil.com/x"
	} else if s := strings.ToLower(m[1]); s == "ftps" || s == "sftp" {
		// ftps and sftp carry a host and path like https does, and the scheme is
		// not part of the pattern, so rewriting is lossless. It is also required:
		// the WHATWG parser treats them as non-special schemes and would hand the
		// Worker an opaque host — not lowercased, not percent-decoded, no IDNA.
		raw = "https://" + raw[len(m[0]):]
	}

	return decodeAuthority(raw)
}

// forbiddenHostBytes are the WHATWG forbidden host code points. '%' is included
// because the host is decoded exactly once, so a surviving '%' is invalid.
const forbiddenHostBytes = "\x00\t\n\r #%/:<>?@[\\]^|"

// decodeAuthority percent-decodes the authority once and reports whether every
// decoded byte is allowed in a host, which is what the WHATWG host parser does.
// net/url instead rejects any escape in the host at all ("invalid URL escape"),
// so without this Go would drop "https://ev%69l.com/x" while the Worker resolved
// it to evil.com — the blocklist would miss an entry the lookup side canonicalizes.
func decodeAuthority(raw string) (string, bool) {
	m := schemeRegexp.FindStringSubmatch(raw)
	if m == nil {
		return raw, true
	}
	start := len(m[0])
	end := len(raw)
	if i := strings.IndexAny(raw[start:], "/?#"); i >= 0 {
		end = start + i
	}
	auth := raw[start:end]
	if !strings.Contains(auth, "%") {
		return raw, true
	}

	var b strings.Builder
	b.Grow(len(auth))
	for i := 0; i < len(auth); i++ {
		if auth[i] == '%' && i+2 < len(auth) && isHex(auth[i+1]) && isHex(auth[i+2]) {
			c := unhex(auth[i+1])<<4 | unhex(auth[i+2])
			if strings.IndexByte(forbiddenHostBytes, c) >= 0 {
				return "", false
			}
			b.WriteByte(c)
			i += 2
			continue
		}
		b.WriteByte(auth[i])
	}
	return raw[:start] + b.String() + raw[end:], true
}

// escapeLonePercent replaces a '%' that does not start a valid escape with
// "%25", which is what the WHATWG parser does. Without it net/url would reject
// the URL while the Worker happily canonicalized it.
func escapeLonePercent(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		if s[i] == '%' && !(i+2 < len(s) && isHex(s[i+1]) && isHex(s[i+2])) {
			b.WriteString("%25")
			continue
		}
		b.WriteByte(s[i])
	}
	return b.String()
}

// canonHost takes the host after preprocess/decodeAuthority has already done the
// single percent-decoding pass, exactly as WHATWG hands one to the Worker.
func canonHost(h string) (string, error) {
	if isUnicode(h) {
		a, err := idn.ToASCII(h)
		if err != nil {
			return "", ErrInvalid
		}
		h = a
	}
	h = dotsRegexp.ReplaceAllString(h, ".")
	h = strings.Trim(h, ".")
	h = strings.ToLower(h)

	if ip := parseIPAddress(h); ip != "" {
		return "", ErrIPHost
	}
	if !hostRegexp.MatchString(h) {
		return "", ErrInvalid
	}
	return h, nil
}

// canonPath decodes the path, resolves "." / ".." / empty segments, then escapes
// exactly once. Equivalent to path.Clean plus Google's trailing-slash rule, but
// written as an explicit loop so the TypeScript port is line-for-line the same.
func canonPath(p string) (string, error) {
	p, err := recursiveUnescape(p)
	if err != nil {
		return "", err
	}
	if !strings.HasPrefix(p, "/") {
		p = "/" + p
	}

	segs := strings.Split(p[1:], "/")
	out := make([]string, 0, len(segs))
	for _, s := range segs {
		switch s {
		case "", ".": // empty and "." segments vanish
		case "..":
			if len(out) > 0 {
				out = out[:len(out)-1]
			}
		default:
			out = append(out, s)
		}
	}

	res := "/" + strings.Join(out, "/")
	if last := segs[len(segs)-1]; len(out) > 0 && (last == "" || last == "." || last == "..") {
		res += "/" // the original ended on a directory
	}
	return escapePath(res), nil
}

File: client/src/canonical.ts
// Canonicalization + candidate expansion for the URL blocklist lookup.
// Mirror of internal/normalize/canonical.go. Both are tested against
// testdata/normalize.json; if they drift, lookups silently stop matching.

import {
  escapePath,
  escapeQuery,
  generateLookupHosts,
  generateLookupPaths,
  parseIPAddress,
  recursiveUnescape,
} from './sb_urls';

export type UrlErrorCode = 'invalid_host' | 'ip_not_allowed';

export class InvalidUrlError extends Error {
  constructor(public readonly code: UrlErrorCode) {
    super(code);
  }
}

/** Must match hostRegexp in canonical.go. */
const HOST_RE = /^(?:[a-z0-9_-]{1,63}\.)+[a-z][a-z0-9-]{0,62}$/;

/** Captures an explicit "scheme://", so "evil.com:8080/x" stays schemeless. */
const SCHEME_RE = /^([a-zA-Z][a-zA-Z0-9+.-]*):\/\//;

export interface Parts {
  host: string;  // lowercase, punycode, no trailing dot
  path: string;  // canonical, always starts with "/"
  query: string; // canonical, without the leading "?"
}

/** The canonical string that gets hashed and stored. */
export function pattern(p: Parts): string {
  return p.query ? p.host + p.path + '?' + p.query : p.host + p.path;
}

/**
 * Canonicalize. Twin of Parse() in canonical.go, and the only entry point —
 * scheme and host policy is enforced by the submit endpoint on the user side and
 * by skipping unusable lines on the feed side.
 *
 * The scheme is dropped from the pattern, so a feed entry "http://evil.com/x"
 * matches a submitted "https://evil.com/x", and an sftp link matches a pattern
 * stored from a web URL on the same host.
 */
export function parse(raw: string): Parts {
  let u: URL;
  try {
    u = new URL(preprocess(raw));
  } catch {
    throw new InvalidUrlError('invalid_host');
  }
  if (u.protocol !== 'http:' && u.protocol !== 'https:') {
    throw new InvalidUrlError('invalid_host');
  }
  if (u.hostname.startsWith('[')) throw new InvalidUrlError('ip_not_allowed'); // IPv6

  return {
    host: canonHost(u.hostname),
    path: canonPath(u.pathname),
    query: escapeQuery(recursiveUnescape(u.search.slice(1))),
  };
}

/**
 * Every pattern that, if present in the blocklist, should block this URL.
 * Up to 7 hosts x 6 paths, deduplicated.
 */
export function candidates(p: Parts): string[] {
  const hosts = generateLookupHosts(p.host, false);
  const paths = generateLookupPaths(p.path, p.query);

  const out = new Set<string>();
  for (const h of hosts) for (const pa of paths) out.add(h + pa);
  return [...out];
}

const enc = new TextEncoder();

/** One indexed D1 query for all candidates. Throws InvalidUrlError on bad input. */
export async function isListed(db: D1Database, rawUrl: string): Promise<boolean> {
  const cands = candidates(parse(rawUrl));
  const hashes = await Promise.all(
    cands.map(async (c) => new Uint8Array(await crypto.subtle.digest('SHA-256', enc.encode(c)))),
  );
  const sql = `SELECT 1 FROM bad_url WHERE h IN (${hashes.map(() => '?').join(',')}) LIMIT 1`;
  return (await db.prepare(sql).bind(...hashes).first()) !== null;
}

// --- internals ---

/** Mirrors preprocess() in canonical.go; the WHATWG parser needs a scheme. */
function preprocess(raw: string): string {
  const hash = raw.indexOf('#');
  if (hash >= 0) raw = raw.slice(0, hash); // drop the fragment
  raw = escapeLonePercent(
    raw
      .trim()
      .replace(/[\t\r\n]/g, '')
      .replace(/\\/g, '/'), // special schemes treat \ as /
  );

  const m = SCHEME_RE.exec(raw);
  if (m === null) return 'http://' + raw; // some feeds ship bare "evil.com/x"

  // ftps and sftp carry a host and path like https does, and the scheme is not
  // part of the pattern, so rewriting is lossless. It is also required: they are
  // non-special schemes, so the WHATWG parser would give them an opaque host —
  // not lowercased, not percent-decoded, no IDNA — that never matches a pattern.
  const scheme = m[1].toLowerCase();
  return scheme === 'ftps' || scheme === 'sftp' ? 'https://' + raw.slice(m[0].length) : raw;
}

/**
 * Replace a '%' that does not start a valid escape with "%25". The WHATWG
 * parser does this anyway; Go needs it explicitly, so both do it here.
 */
function escapeLonePercent(s: string): string {
  let out = '';
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '%' && !(i + 2 < s.length && isHexChar(s[i + 1]) && isHexChar(s[i + 2]))) {
      out += '%25';
      continue;
    }
    out += s[i];
  }
  return out;
}

function isHexChar(c: string): boolean {
  return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}

/**
 * WHATWG has already lowercased the host, percent-decoded it once, applied IDNA
 * and thrown on forbidden host code points. Go reproduces all of that in
 * preprocess/decodeAuthority, because net/url rejects escapes in the host.
 */
function canonHost(h: string): string {
  h = h.replace(/[.]+/g, '.').replace(/^\.+|\.+$/g, '').toLowerCase();

  if (parseIPAddress(h) !== '') throw new InvalidUrlError('ip_not_allowed');
  if (!HOST_RE.test(h)) throw new InvalidUrlError('invalid_host');
  return h;
}

/**
 * Decode the path, resolve "." / ".." / empty segments, then escape once.
 * The WHATWG parser resolves dot segments too, but only the ones visible before
 * decoding and without collapsing "//", so this runs regardless.
 */
function canonPath(p: string): string {
  p = recursiveUnescape(p);
  if (!p.startsWith('/')) p = '/' + p;

  const segs = p.slice(1).split('/');
  const out: string[] = [];
  for (const s of segs) {
    if (s === '' || s === '.') continue;
    if (s === '..') {
      out.pop();
      continue;
    }
    out.push(s);
  }

  let res = '/' + out.join('/');
  const last = segs[segs.length - 1];
  if (out.length > 0 && (last === '' || last === '.' || last === '..')) res += '/';
  return escapePath(res);
}
